docs(blog): add Refine and Material UI usage docs (#2199)
pankod/refine: documentation/blog/2022-07-21-admin-panel-with-materialui-and-strapi.md

---

> title: Building a Refine CRUD app with Material UI and Strapi 

Reviewer: ```suggestion
title: Building a CRUD app with Material UI and Strapi 
```

> description: test

Reviewer: ```suggestion
description: test
```
...

slug: build-admin-panel-with-material-ui-and-strapi
authors: necati

> tags: [refine, fundamentals, react, javascript, low-code, internal-tools]

Reviewer: ```suggestion
tags: [refine, fundamentals, react, javascript, low-code, internal-tools, crud, material-ui, mui]
```

---
title: Building a CRUD app with Material UI and Strapi 
description: How to build CRUD admin panel with Material UI?
slug: build-admin-panel-with-material-ui-and-strapi
authors: necati
tags: [refine, fundamentals, react, javascript, low-code, internal-tools, crud, material-ui, mui]
image: https://refine.dev/img/refine_social.png
hide_table_of_contents: false
---

import listpage from '@site/static/img/blog/2022-07-21-admin-panel-with-materialui-and-strapi/list-page.png';
import createGif from '@site/static/img/blog/2022-07-21-admin-panel-with-materialui-and-strapi/create.gif'
import deleteGif from '@site/static/img/blog/2022-07-21-admin-panel-with-materialui-and-strapi/delete-record.gif'
import undoableGif from '@site/static/img/blog/2022-07-21-admin-panel-with-materialui-and-strapi/undoable-mode.gif'

## Introduction

> We will build an admin panel that supports CRUD operations, has built in authentication, and a mutation mode feature using industry-standard best tools.

Reviewer: ```suggestion
We will build an **admin panel** that supports **CRUD** operations, has built-in **authentication**, and a mutation mode ([?](https://refine.dev/docs/guides-and-concepts/mutation-mode/))  feature using industry-standard best tools.
```

> UI design can be a complex and time-consuming process, but using a tool like Material UI can help to simplify the process and speed up the development cycle. In this tutorial we'll use the benefits of Material UI and to handle data fetching and mutations. We'll also integrate StrapiV4 data provider that Refine has built-in support.

Reviewer: ```suggestion
UI design can be a complex and time-consuming process, but a tool like [Material UI](https://mui.com) can help simplify the process and speed up the development cycle. In this tutorial, we'll use Material UI's benefits and handle data fetching and mutations. We'll also integrate the [Strapi](https://strapi.io/) data provider that Refine has built-in support.
```

We'll walk through the process of listing, creating and deleting posts in a Refine application and make use of Refine's components and hooks to build out our functionality.

Steps we'll cover includes:
- What are the benefits of using Refine?
- Bootstrapping the Refine App

> - Implementing Strapi-v4 data provider

Reviewer: ```suggestion
- Implementing Strapi data provider
```

- Creating a list page

> - Handling relational data

Reviewer: ```suggestion
- Handling relational data with [Strapi Population support](https://docs.strapi.io/developer-docs/latest/developer-resources/database-apis-reference/rest/populating-fields.html)
```

- CRUD operations
- Implementing mutation mode
- Sharing the current page and location easily using `syncwithlocation` feature.
<!--truncate-->

## Prerequisities
Before we dive into the meat of the article, let's first take a look at the tools documents we'll be using. 
- [Refine](https://refine.dev/docs/getting-started/overview/)
- [Refine StrapiV4 data provider ](https://refine.dev/docs/examples/data-provider/strapi-v4/)
- [Material UI](https://mui.com/material-ui/getting-started/overview/)
- [Refine Material UI Tutorial](https://refine.dev/docs/ui-frameworks/mui/tutorial/)

##  What are the benefits of using Refine?

Refine is a headless React internal tool framework that helps you develop quickly while developing both B2B and B2C applications. It speeds you up while allowing full customization, making it an ideal choice for rapid development with pro features.

-   It is Open Source under the MIT license.
-   It is easy to use and learn. There are many examples to help you get started, as well as documentation.
-   Is a framework that does not require you to use any UI libraries or frameworks.

> -   Supports Ant Design and Material UI for quick and easy solutions.

Reviewer: ```suggestion
-   Supports Ant Design and Material UI natively for quick and easy solutions. Thanks to the headless approach, you can easily integrate your own UI solution.
```

-   Backend agnostic, so you can connect to any backend you want.
-   Customizable, which means you can change it to fit your needs.
-   Some of the main features are data fetching and state management, routings, authentication, authorization, internationalization, real-time, mutation modes with optimistic and pessimistic and undoable modes

## Bootstrapping the Refine app
We'll use [superplate](https://github.com/pankod/superplate) CLI wizard to create and customize Refine application.

Run the following command
```
npx superplate-cli -p refine-react materialui-example
```

Select the following options to complete CLI wizard:

```
? Do you want to use a UI Framework?:
❯ Material UI

? Do you want an extended theme?:
❯ No

? Do you want to add dark mode support?:
❯ No

? Router Provider:
❯ React Router v6

? Data Provider:
❯ Strapi v4

? Do you want a customized layout?
❯ No

? i18n - Internationalization:
❯ No
```

CLI should be create a project and install the selected dependencies.

### Implementing Strapi-v4 data provider
Data providers are refine components making it possible to consume different API's and data services conveniently.

> The required Strapi-V4 data provider setups are added automatically by CLI wizard.

Reviewer: ```suggestion
The required Strapi data provider setups are added automatically by the CLI wizard.
```

> To consume Refine's Fake StrapiV4 data provider, we'll need change API URL in the project folder.

Reviewer: ```suggestion
To consume Refine's Fake Strapi API, we'll need to change the `API URL` in the project folder.
```

Imagine we need to share the current page with filtering and sorting parameters to our collegues. The proper way to do is, sharing the URL that has include all needed parameters like:

```
/posts?current=1&pageSize=8&sort[]=createdAt&order[]=desc
```

Refine offers `syncWithLocation` property that allow us to editing query parameters manually and share current page, items count per page, sort and filter parameters easily to others.

```tsx title="src/App.tsx"
...
function App() {
    return (
        <ThemeProvider theme={LightTheme}>
            <CssBaseline />
            <GlobalStyles styles={{ html: { WebkitFontSmoothing: "auto" } }} />
            <RefineSnackbarProvider>
                <Refine
                    ...
                    mutationMode="undoable"
                    //highlight-next-line
                    syncWithLocation
                />
            </RefineSnackbarProvider>
        </ThemeProvider>
    );
}

export default App;
```

Now, we can get current information from URL as a query parameters. We can either use this link to share to others or define filter, pagination, and sorting parameters manually from changing URL parameters.

 ## Conclusion

> In this article, we'll show you how to build a CRUD admin panel using refine and Material UI. This approach will allow you to quickly create an admin interface for your application with minimal coding. We'll start by setting up our project with the required dependencies. Then, we'll create our CRUD components using Material UI. Finally, we'll wire everything up and add some extra features from refine like mutation mode. 
> 

Reviewer: ```suggestion
In this article, we'll show you how to build a **CRUD admin panel** using refine and **Material UI**. This approach will allow you to quickly create an admin interface for your application with minimal coding. We'll start by setting up our project with the required dependencies. Then, we'll create our CRUD components using Material UI. Finally, we'll wire everything up and add some extra features from refine like mutation mode. 

```

We covered:
- How to bootstrap refine app
- Connecting Strapiv4 data provider to refine app.
- Creating pages for CRUD operations
- Implementing some of refine features like mutation mode and location sync.

refine is an open source tool that rapidly and flexibly develops for CRUD admin panels or web apps. It is easy to get started with and doesn't require a lot of code. It has nice documentation that covered examples, guidelines, and tutorials using best practices. refine is constantly being updated with new features and improvements.

[Refer to official refine page for more information &#8594](https://refine.dev/)
  

> 

Reviewer: here we can dynamically add sandbox from StackBlitz

Steps we'll cover includes:
- What are the benefits of using Refine?
- Bootstrapping the Refine App
- Implementing Strapi-v4 data provider
- Creating page
- CRUD operations
- Adding authentication
- Implementing mutation mode
- Testing syncwithlocation feature
<!--truncate-->

## Prerequisities
Before we dive into the meat of the article, let's first take a look at the tools we'll be using. 
- [Refine](https://refine.dev/docs/getting-started/overview/)
- [Refine StrapiV4 data provider ](https://refine.dev/docs/examples/data-provider/strapi-v4/)
- [Material UI](https://mui.com/material-ui/getting-started/overview/)
- [Refine Material UI Tutorial](https://refine.dev/docs/ui-frameworks/mui/tutorial/)

##  What are the benefits of using Refine?

Refine is a headless React internal tool framework that helps you develop quickly while developing both B2B and B2C applications. t speeds you up while allowing full customization, making it an ideal choice for rapid development with pro features.

-   It is Open Source under the MIT license.

> -   Easy to use and it is easy to learn. There are many examples to help you get started, as well as documentation.

Reviewer 2: ```suggestion
-   It is easy to use and learn. There are many examples to help you get started, as well as documentation.
```

-   Is a framework that does not require you to use any UI libraries or frameworks.
-   Supports Ant Design and Material UI for quick and easy solutions.
-   Backend agnostic, so you can connect to any backend you want.
-   Customizable, which means you can change it to fit your needs.

> -   Some of main features are data fetching and state management, routings, authentication, authorization, internationalization, real-time, mutation modes with optimistic and pessimistic and undoadable modes

Reviewer 2: ```suggestion
-   Some of the main features are data fetching and state management, routings, authentication, authorization, internationalization, real-time, mutation modes with optimistic and pessimistic and undoable modes
```

                field: "actions",
                minWidth: 180,
                align: "center",
                flex: 1,
                renderCell: function render(params) {
                    return (
                        <Stack direction="row" spacing={1}>
                            <EditButton
                                size="small"
                                hideText
                                recordItemId={params.row.id}
                            />
                            <ShowButton
                                size="small"
                                hideText
                                recordItemId={params.row.id}
                            />
                            <DeleteButton
                                size="small"
                                hideText
                                recordItemId={params.row.id}
                            />
                        </Stack>
                    );
                },
            },
        ],
        [],
    );

    return (
        <List>
            <DataGrid {...dataGridProps} columns={columns} autoHeight />
        </List>
    );
};

```

> We use Material UI components that exports from Refine.

Reviewer 2: ```suggestion
We use Material UI components exports from Refine.
```

[Get more information about components and hooks in PostList page we defined&#8594](https://refine.dev/docs/ui-frameworks/mui/tutorial/#creating-a-list-page)

[Refer to Material UI offical docs to learn more about components&#8594](https://mui.com/x/react-data-grid/components/#main-content)

> Note you will need a few more files which help src/App.tsx to find your pages and posts. In the /pages folder, put this index.tsx file in it which allows everything in the posts folder to be used elsewhere.

Reviewer 2: ```suggestion
Note you will need a few more files that help src/App.tsx to find your pages and posts. In the /pages folder, put this index.tsx file in it which allows everything in the posts folder to be used elsewhere.
```
